### PR TITLE
Problem: cannot generate ledger address (fix #326)

### LINKED ISSUE #326: Problem: cannot generate ledger address
ethermintd  keys add my --ledger
```
Error: address  rejected
```

### PR DESCRIPTION
**Pull request by leejw51crypto**
Solution: add ledger command

1. ethermintd ledger add
2. ethermintd ledger send 0x...from   0x...to 2

### CODE REVIEW

**Comment by fedekunze** [comment]
Thanks, @leejw51crypto, but the PR doesn't build. Can you fix that?

**Comment by leejw51crypto** [comment]
ok, i'll check

**Review by thomas-nguy — APPROVED** [review]
LGTM
Note : another to do would be to use an http client and call `eth` api to get nonce, gas estimation and send transactions. That way you won't have to duplicate the code

File: cmd/ethermintd/ledger.go [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	closeWalletErr := firstLedgerWallet.Close()
+	if closeWalletErr != nil {
+		return fmt.Errorf("ledger close error %v", closeWalletErr)
+	}
+	return nil
+}
+
+func ledgerCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:                        "ledger",
```
Comment (fedekunze) [inline cmd/ethermintd/ledger.go:57]: I don't understand why are we creating a new `"ledger"` command... Why can't we create a custom `keys add --ledger` instead?

File: cmd/ethermintd/ledger.go [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+	"github.com/spf13/cobra"
+	"github.com/tendermint/tendermint/libs/log"
+	rpctypes "github.com/tharsis/ethermint/ethereum/rpc/types"
+)
+
+const (
+	flagIndex = "index"
+)
+
+func runAddCmd(cmd *cobra.Command, _ []string) error {
```
Comment (fedekunze) [inline cmd/ethermintd/ledger.go:25]: is this stored in the keyring?

File: cmd/ethermintd/ledger.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	allLedgerWallets := ledgerHub.Wallets()
+	firstLedgerWallet := allLedgerWallets[0]
+	openWalletErr := firstLedgerWallet.Open("")
+	if openWalletErr != nil {
+		return fmt.Errorf("ledger open error %vn", openWalletErr)
+
+	}
+
+	// bip44(44), coin type(60), account, change ,index
+	hdPath := []uint32{0x80000000 + 44, 0x80000000 + 60, 0x80000000 + 0, 0, bip44Index}
```
Comment (fedekunze) [inline cmd/ethermintd/ledger.go:40]: use go-ethereum variables

File: cmd/ethermintd/ledger.go [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+	derivedAddress, deriveErr := firstLedgerWallet.Derive(hdPath, true)
+	if deriveErr != nil {
+		return fmt.Errorf("derive address error %v", deriveErr)
+	}
+
+	bech32Address, _ := sdk.Bech32ifyAddressBytes(sdk.GetConfig().GetBech32AccountAddrPrefix(), derivedAddress.Address[:])
+	fmt.Printf("Ledger Address Index= %d\nAddress= %s  %s\n", bip44Index, derivedAddress.Address.Hex(), bech32Address)
+	closeWalletErr := firstLedgerWallet.Close()
```
Comment (fedekunze) [inline cmd/ethermintd/ledger.go:48]: this should be called through `defer()`

File: cmd/ethermintd/ledger.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+	addCmd := &cobra.Command{
+		Use:   "add",
+		Short: "Add ledger address",
+		RunE:  runAddCmd,
+	}
+	addCmd.Flags().Uint32(flagIndex, 0, "Address index number for HD derivation")
+
+	// send
+	sendCmd := &cobra.Command{
+		Use: "send [from_key_or_address] [to_address] [amount]",
```
Comment (fedekunze) [inline cmd/ethermintd/ledger.go:75]: this is unrelated to the PR description (adding an address / key through ledger)

File: cmd/ethermintd/root.go [hunk 6]
```diff
@@ -114,6 +114,7 @@ func NewRootCmd() (*cobra.Command, params.EncodingConfig) {
 		rpc.StatusCommand(),
 		queryCommand(),
 		txCommand(),
+		ledgerCommand(),
```
Comment (fedekunze) [inline cmd/ethermintd/root.go:117]: we should remove this and just fix the `keys add --ledger` command

**Review by fedekunze — CHANGES_REQUESTED** [review]
I'm not a super fan of the approach from this PR (see comments). The `send` command is also out of scope and can be performed using the ethereum ledger app already

File: cmd/ethermintd/ledger.go (hunk 1 above)
Comment (leejw51crypto) [reply cmd/ethermintd/ledger.go:57]: current cosmos ledger only working with cosmos ledger app , cannot use override

File: cmd/ethermintd/ledger.go (hunk 2 above)
Comment (leejw51crypto) [reply cmd/ethermintd/ledger.go:25]: no further information is not stored in the keyring for ledger

File: cmd/ethermintd/ledger.go [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+	// bip44(44), coin type(60), account, change ,index
+	hdPath := []uint32{0x80000000 + 44, 0x80000000 + 60, 0x80000000 + 0, 0, bip44Index}
+	derivedAddress, deriveErr := firstLedgerWallet.Derive(hdPath, true)
+	if deriveErr != nil {
+		return fmt.Errorf("derive address error %v", deriveErr)
+	}
+
+	bech32Address, _ := sdk.Bech32ifyAddressBytes(sdk.GetConfig().GetBech32AccountAddrPrefix(), derivedAddress.Address[:])
+	fmt.Printf("Ledger Address Index= %d\nAddress= %s  %s\n", bip44Index, derivedAddress.Address.Hex(), bech32Address)
```
[the remark this answers is not in the document]
Comment (leejw51crypto) [reply cmd/ethermintd/ledger.go:47]: it's for ledger address information display.

File: cmd/ethermintd/ledger.go (hunk 4 above)
Comment (leejw51crypto) [reply cmd/ethermintd/ledger.go:48]: ok

File: cmd/ethermintd/ledger.go (hunk 5 above)
Comment (leejw51crypto) [reply cmd/ethermintd/ledger.go:75]: this is for generation ,  signing together.
it's very tightly coupled

File: cmd/ethermintd/root.go (hunk 6 above)
Comment (leejw51crypto) [reply cmd/ethermintd/root.go:117]: i  inspected cosmos level modification, that requires overriding whole cosmos-sdk.
but cosmo-sdk ledger is not possible, because eth ledger app only working with eth style signing.

File: cmd/ethermintd/ledger.go (hunk 1 above)
Comment (leejw51crypto) [reply cmd/ethermintd/ledger.go:57]: so in this pr, cosmos ledger is just turned off , and added eth style ledger.
this is very simpler, no side effects for cosmos codes.

File: cmd/ethermintd/ledger.go (hunk 3 above)
Comment (leejw51crypto) [reply cmd/ethermintd/ledger.go:40]: ok

File: cmd/ethermintd/root.go (hunk 6 above)
Comment (leejw51crypto) [reply cmd/ethermintd/root.go:117]: another solution is using

https://github.com/LedgerHQ/ledgerjs/tree/master/packages/hw-app-eth#signeip712hashedmessage

just signing with hash?
then we can do it in cosmos-level.

- good: supporting all
- bad: signer cannot see contents of the message to sign

Comment (leejw51crypto) [reply cmd/ethermintd/root.go:117]: i'll check signing hash

**Comment by fedekunze** [comment]
> bad: signer cannot see contents of the message to sign

Based on this, we decided this is a security issue we don't want to deal with. This is cons outweighs the benefit of this PR
